### ISSUE TITLE
tensorflow serving returning NaN when predict

### ISSUE THREAD

**Issue by zhaocc1106** [tensorflow/serving#1670]
Tensorflow version: 2.2.0
Tensorflow serving version: TensorFlow ModelServer: 2.2.0-rc2+dev.sha.d22fc19 TensorFlow Library: 2.2.0

I had trained one GAN model and saved the generator by the following function:
```
    tf.keras.models.save_model(
        generator,
        filepath=os.path.join(MODEL_PATH, 'model_saver'),
        overwrite=True,
        include_optimizer=False,
        save_format=None,
        options=None
    )
```

It predicts successfully when load model by tf.keras.models.load_model in python. But when serving the model in tensorflow model server, the model returns NaN value. I serve the model as follows:
```
zhaocc:~/products/tensorflow_server$ sudo docker run -t --rm -p 8502:8501     -v "/tmp/pix2pix/sketch_photo/model_saver:/models/photo2sketch"     -e MODEL_NAME=photo2sketch     tensorflow/serving &
[3] 30089
zhaocc:~/products/tensorflow_server$ 2020-06-17 12:57:31.745339: I tensorflow_serving/model_servers/server.cc:86] Building single TensorFlow model file config:  model_name: photo2sketch model_base_path: /models/photo2sketch
2020-06-17 12:57:31.745448: I tensorflow_serving/model_servers/server_core.cc:464] Adding/updating models.
2020-06-17 12:57:31.745459: I tensorflow_serving/model_servers/server_core.cc:575]  (Re-)adding model: photo2sketch
2020-06-17 12:57:31.846162: I tensorflow_serving/core/basic_manager.cc:739] Successfully reserved resources to load servable {name: photo2sketch version: 1}
2020-06-17 12:57:31.846213: I tensorflow_serving/core/loader_harness.cc:66] Approving load for servable version {name: photo2sketch version: 1}
2020-06-17 12:57:31.846233: I tensorflow_serving/core/loader_harness.cc:74] Loading servable version {name: photo2sketch version: 1}
2020-06-17 12:57:31.846282: I external/org_tensorflow/tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /models/photo2sketch/1
2020-06-17 12:57:31.874158: I external/org_tensorflow/tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { serve }
2020-06-17 12:57:31.874182: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:295] Reading SavedModel debug info (if present) from: /models/photo2sketch/1
2020-06-17 12:57:31.874315: I external/org_tensorflow/tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-17 12:57:31.952982: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:234] Restoring SavedModel bundle.
2020-06-17 12:57:32.172641: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:183] Running initialization op on SavedModel bundle at path: /models/photo2sketch/1
2020-06-17 12:57:32.248514: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:364] SavedModel load for tags { serve }; Status: success: OK. Took 402236 microseconds.
2020-06-17 12:57:32.256576: I tensorflow_serving/servables/tensorflow/saved_model_warmup.cc:105] No warmup data file found at /models/photo2sketch/1/assets.extra/tf_serving_warmup_requests
2020-06-17 12:57:32.265064: I tensorflow_serving/core/loader_harness.cc:87] Successfully loaded servable version {name: photo2sketch version: 1}
2020-06-17 12:57:32.267113: I tensorflow_serving/model_servers/server.cc:355] Running gRPC ModelServer at 0.0.0.0:8500 ...
[warn] getaddrinfo: address family for nodename not supported
2020-06-17 12:57:32.269289: I tensorflow_serving/model_servers/server.cc:375] Exporting HTTP/REST API at:localhost:8501 ...
[evhttp_server.cc : 238] NET_LOG: Entering the event loop ...
```

When I predict by REST request, it return NaN with correct shape:

```
[[[[nan nan nan]
   [nan nan nan]
   [nan nan nan]
   ...
   [nan nan nan]
   [nan nan nan]
   [nan nan nan]]
```

The generator functional model is as follows:
```
def downsample(filters, size, apply_batchnorm=True):
    """Use convolution layer to downsample

    Args:
        filters: The filters number.
        size: The kernel size.
        apply_batchnorm: If apply batch normalization.

    Returns:
        The downsample model.
    """
    initializer = tf.random_normal_initializer(0, 0.02)

    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Conv2D(filters=filters,
                                     kernel_size=(size, size),
                                     strides=(2, 2),
                                     padding='same',
                                     kernel_initializer=initializer,
                                     use_bias=False))
    if apply_batchnorm:
        model.add(tf.keras.layers.BatchNormalization())
    model.add(tf.keras.layers.LeakyReLU())
    return model


def upsample(filters, size, apply_dropout=True):
    """Use convolution layers to upsample.

    Args:
        filters: The filters number.
        size: The kernel size.
        apply_dropout: If apply dropout.

    Returns:
        The upsample model.
    """
    initializer = tf.random_normal_initializer(0, 0.02)

    model = tf.keras.models.Sequential()
    model.add(tf.keras.layers.Conv2DTranspose(filters=filters,
                                              kernel_size=(size, size),
                                              strides=(2, 2),
                                              padding="same",
                                              kernel_initializer=initializer,
                                              use_bias=False))
    model.add(tf.keras.layers.BatchNormalization())

    if apply_dropout:
        model.add(tf.keras.layers.Dropout(0.5))
    model.add(tf.keras.layers.ReLU())
    return model


def Generator():
    """The generator model."""
    inputs = tf.keras.layers.Input(shape=[IMG_HEIGHT, IMG_WIDTH, 3])
    down_stack = [
        downsample(64, 4, apply_batchnorm=True),  # (bs, 128, 128, 64)
        downsample(128, 4),  # (bs, 64, 64, 128)
        downsample(256, 4),  # (bs, 32, 32, 256)
        downsample(512, 4),  # (bs, 16, 16, 512)
        downsample(512, 4),  # (bs, 8, 8, 512)
        downsample(512, 4),  # (bs, 4, 4, 512)
        downsample(512, 4),  # (bs, 2, 2, 512)
        downsample(512, 4),  # (bs, 1, 1, 512)
    ]

    up_stack = [
        upsample(512, 4, apply_dropout=True),  # (bs, 2, 2, 1024)
        upsample(512, 4, apply_dropout=True),  # (bs, 4, 4, 1024)
        upsample(512, 4, apply_dropout=True),  # (bs, 8, 8, 1024)
        upsample(512, 4),  # (bs, 16, 16, 1024)
        upsample(256, 4),  # (bs, 32, 32, 512)
        upsample(128, 4),  # (bs, 64, 64, 256)
        upsample(64, 4),  # (bs, 128, 128, 128)
    ]

    initializer = tf.random_normal_initializer(0., 0.02)
    last = tf.keras.layers.Conv2DTranspose(OUTPUT_CHANNELS, 4,
                                           strides=2,
                                           padding='same',
                                           kernel_initializer=initializer,
                                           activation='tanh')  # (bs, 256, 256, 3)

    x = inputs

    # Downsampling through the model.
    skips = []
    for down in down_stack:
        x = down(x)
        skips.append(x)

    skips = reversed(skips[:-1])

    # Upsampling through the model.
    for (up, skip) in zip(up_stack, skips):
        x = up(x)
        x = tf.keras.layers.Concatenate()([x, skip])  # Skip connect.

    x = last(x)
    return tf.keras.Model(inputs=inputs, outputs=x)
```

The complete code link is as follows:
https://github.com/zhaocc1106/machine_learn/blob/master/NeuralNetworks-tensorflow/generation_network_model/GAN/photo2sketch.py
How can I debug it? Thanks very much!

**Comment by zhaocc1106** [comment]
I tried serve this model on gpu-serving. Has the same problem.
Serving another rnn classifer model is okay...

**Comment by rmothukuru** [comment]
@zhaocc1106,
Can you please provide the `SignatureDef` of your `Model` (using `saved_model_cli` command) and the Command that you have used to perform Inference using `REST API` so that we can reproduce the issue? Thanks!

**Comment by zhaocc1106** [comment]
Like the follows:
https://stackoverflow.com/a/62483768/13766029
There is  temporary workaround solution and the problem maybe caused by 'training' param.

**Comment by gowthamkpr** [comment]
@zhaocc1106 Can you please explain in depth about the training param? Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than 2 weeks. Please add additional comments to open this issue again.Thanks!

**Comment by taninaim** [comment]
I'm having the same problem. I served an image classifier and it's giving me nan on only some images. I stacked 4 images so I have a tensor of shape (4, 380, 380, 3). Now the prediction worked fine for the first image, but the next three returned nans. I've checked that the values are properly normalized. This is very weird because I've manually loaded the model using tf.keras.models.load_model and it worked totally fine on the same images that returned nan from tf-serving.

**Comment by albert-92** [comment]
@taninaim I am having the exact same issue. For some images it works fine and for others it return nans.
Do you use EfficientNet? Did you find any solution for that problem?

**Comment by zhaocc1106** [comment]
> @taninaim I am having the exact same issue. For some images it works fine and for others it return nans.
> Do you use EfficientNet? Did you find any solution for that problem?

Like the follows:
https://stackoverflow.com/a/62483768/13766029
I set the 'training' param to false, which solve my problem. I think it's a bug.

**Comment by UlionTse** [comment]
Same Question.

**Comment by Pamulapati13** [comment]
I too have the same issue, the model has two outputs, while training on Google Colab it works fine, but when I run on my machine locally it returns 0's for one branch and NaNs for the other branch.

**Comment by shridharvarun** [comment]
Having a similar issue. No way of providing `training=True` param while predicting using tfx

**Comment by satyaki-mallick** [comment]
Same issue as well. I am trying to predict through REST in AWS Sagemaker Tensorflow Estimator.

**Comment by omair-kg** [comment]
During inference, training should ideally be False i.e. batch norm parameters should be frozen and dropout should be off. When I set training=True in this particular case, i do get a non nan answer but is this answer valid?

Also any explanation for this behaviour when training is set to False.
